### ISSUE TITLE
Mutations do not respect select permissions of a relationship

### ISSUE THREAD

**Issue by CTOJoe** [hasura/graphql-engine#3378]
As a new-comer to the Hasura GraphQL engine, I am very impressed with the functionality and performance of the project. I have been taking a deep dive into how we could utilize this project and potentially replace our current REST API backend.

While experimenting, I found it unexpected that mutations did not respect the select permissions of foreign key relationships. 

Take the following basic schema for consideration
```
Account {
    id: uuid!
    name: String!
    plan_id: uuid!
    plan: Plan!
}
```

```
Plan {
    id: uuid!
    name: String!
    public: Boolean!
}
```

And the following foreign key
```
Account.plan_id -> Plan.id
```

The role of `user` has a row select permission for `Plans` where `public` equals `true`.

When querying `Plans` using the role of `user`, only plans that have `public` set to `true` are returned.

When mutating an `Account` and setting the `plan_id`, the `Plan` select permissions are not taken into consideration. This allows the `user` role to set the `plan_id` to one that the `user` role does not have permission to see. In this case, they would be able to set the `plan_id` to a plan where `public` is not `true`. 

The `returning` field after the mutation does not contain any data on the `Plan` that is not `public`. However, the underlying `plan_id` is set to the one submitted using `user` role.

Due to the foreign key constraint, this would allow the `user` role to discover the `id` of a `Plan` that they should not be able to see or have knowledge of existence. Additionally, this potentially could become a way to connect to nodes outside of an expected select permission scope in a multi-tenant app.

Thanks in advance for any feedback!

**Comment by rakeshkky** [comment]
Hi @CTOJoe, Thank you for trying out Hasura :slightly_smiling_face: 

> When mutating an Account and setting the plan_id, the Plan select permissions are not taken into consideration. This allows the user role to set the plan_id to one that the user role does not have permission to see. In this case, they would be able to set the plan_id to a plan where public is not true.

You've to explicitly define the `update` permissions on `Account` to only allow `user` to set any `plan_id` where `Plan.public` is `true`. 

I'm using the following Postgres schema to replicate your use case (using `integer` for `id`)

```plpgsql
create table plan (id serial primary key, name text not null, public boolean);
create table account (id serial primary key, name text not null, plan_id integer not null references plan(id) );
```
After tracking tables and creating required relationships, the update permission defined as following

![2019-11-19-175115_614x809](https://user-images.githubusercontent.com/12475069/69146191-4ccd2700-0af5-11ea-9ee8-c6207c5b6e33.png)

**Comment by CTOJoe** [comment]
Thanks for the quick response. Unfortunately the rule only denies the role of `user` from updating an `account` where the related `plan` has `public` equal to `true`. This does not prevent the `user` role from setting the `plan_id` to one that is outside their select permission scope.

To re-produce the unexpected outcome I am talking about, please start with an empty database and run the following sql
```sql
CREATE TABLE plans ( id SERIAL PRIMARY KEY, name TEXT NOT NULL, public BOOLEAN NOT NULL );
CREATE TABLE accounts ( id SERIAL PRIMARY KEY, name TEXT NOT NULL, plan_id INTEGER NOT NULL REFERENCES plans(id) );

INSERT INTO plans (name, public) VALUES ('Public Plan', true), ('Private Plan', false);
INSERT INTO accounts (name, plan_id) VALUES  ('Account 1', 1), ('Account 2', 1);
```

After creating the base dataset, please import the following metadata
```javascript
{
    "functions": [],
    "remote_schemas": [],
    "query_collections": [],
    "allowlist": [],
    "tables": [{
        "table": "accounts",
        "is_enum": false,
        "configuration": {
            "custom_root_fields": {
                "select": null,
                "select_by_pk": null,
                "select_aggregate": null,
                "insert": null,
                "update": null,
                "delete": null
            },
            "custom_column_names": {}
        },
        "object_relationships": [{
            "using": {
                "foreign_key_constraint_on": "plan_id"
            },
            "name": "plan",
            "comment": null
        }],
        "array_relationships": [],
        "insert_permissions": [],
        "select_permissions": [{
            "role": "user",
            "comment": null,
            "permission": {
                "allow_aggregations": false,
                "computed_fields": [],
                "columns": ["id", "name", "plan_id"],
                "filter": {
                    "id": {
                        "_eq": "X-Hasura-Account-Id"
                    }
                }
            }
        }],
        "update_permissions": [{
            "role": "user",
            "comment": null,
            "permission": {
                "set": {},
                "columns": ["name", "plan_id"],
                "filter": {
                    "_and": [{
                        "id": {
                            "_eq": "X-Hasura-Account-Id"
                        }
                    }, {
                        "plan": {
                            "public": {
                                "_eq": true
                            }
                        }
                    }]
                }
            }
        }],
        "delete_permissions": [],
        "event_triggers": [],
        "computed_fields": []
    }, {
        "table": "plans",
        "is_enum": false,
        "configuration": {
            "custom_root_fields": {
                "select": null,
                "select_by_pk": null,
                "select_aggregate": null,
                "insert": null,
                "update": null,
                "delete": null
            },
            "custom_column_names": {}
        },
        "object_relationships": [],
        "array_relationships": [{
            "using": {
                "foreign_key_constraint_on": {
                    "column": "plan_id",
                    "table": "accounts"
                }
            },
            "name": "accounts",
            "comment": null
        }],
        "insert_permissions": [],
        "select_permissions": [{
            "role": "user",
            "comment": null,
            "permission": {
                "allow_aggregations": false,
                "computed_fields": [],
                "columns": ["id", "name", "public"],
                "filter": {
                    "public": {
                        "_eq": true
                    }
                }
            }
        }],
        "update_permissions": [],
        "delete_permissions": [],
        "event_triggers": [],
        "computed_fields": []
    }]
}
```

Once the database is initialized and the meta data is imported, click on the GraphQL tab and set a the request headers of `x-hasura-role` to `user` `x-hasura-account-id` to `1`

Paste the following into GraphiQL
```graphql
mutation SetToPrivate {
  update_accounts(where: {}, _set: {
    plan_id: 2
  }) {
    affected_rows
  }
}

mutation SetToPublic {
  update_accounts(where: {}, _set: {
    plan_id: 1
  }) {
    affected_rows
  }
}

query ShowAccount {
  accounts {
    name
    plan {
      id
      public
    }
  }
}

query ListPlans {
  plans {
    id
    public
  }
}
```

Run the `ListPlans` query. Because the role is set to `user` only the public plans are listed. The result is as follows.
```javascript
{
  "data": {
    "plans": [
      {
        "id": 1,
        "public": true
      }
    ]
  }
}
```

Now run the `ShowAccount` query. Because the `user` role is scoped using the session variable `x-hasura-account-id`, they result is only for single account that matches this declared id.
```javascript
{
  "data": {
    "accounts": [
      {
        "name": "Account 1",
        "plan": {
          "id": 1,
          "public": true
        }
      }
    ]
  }
}
```

Now run the `SetToPrivate` mutation. This sets the `plan_id` to an unknown `plan_id` with `public` set as `false`. The result is as follows:
```javascript
{
  "data": {
    "update_accounts": {
      "affected_rows": 1
    }
  }
}
```

If we run the `ShowAccount` query. We now get the following:
```javascript
{
  "data": {
    "accounts": [
      {
        "name": "Account 1",
        "plan": null
      }
    ]
  }
}
```

The `accounts` update rules for the role of `user` requires `id` to be equal to the session variable `x-hasura-account-id` and its related `plan` property of `public` to be equal to `true` as advised. If we run the `SetToPublic` mutation no rows are affected because the account no longer has a `plan` with `public` set to `true`. This can be seen below
```javascript
{
  "data": {
    "update_accounts": {
      "affected_rows": 0
    }
  }
}
```

The role of `user` was successfully able to set a value to an `id` outside of the expected select permissions. Furthermore, they were able to determine that a `plan` with an `id` of `2` exists because the foreign key constraint did not reject the mutation. 

I thought the expected behavior would be a pre-query check for foreign keys, validating the `id` is within the permission scope of what the role can select on the foreign table. Because there is no check, a role could potentially "stitch" into unexpected entities by guessing an `id` that is not within the foreign table select permission scope. In multi-tenant apps, this could potentially be a method of retrieving entities that belong to another tenant.

Let me your thoughts. Thanks!

**Comment by 0x777** [comment]
Hi @CTOJoe, thanks for the detailed explanation. Our current update permission system lacks this [feature](https://github.com/hasura/graphql-engine/issues/384), i.e, you cannot specify a constraint (like the `check` constraint in insert permissions) on the **updated** data. Our recommendation till we add this feature is to disallow updating such columns or write a Postgres trigger which would reject such updates: you can access the current session variables inside the Postgres trigger using `current_setting('hasura.user')::json`.

**Comment by CTOJoe** [comment]
Thank you very much @0x777 for the response. We will look into writing custom triggers and utilizing access to the current session variables. I will close this issue and subscribe to the feature request you mentioned above.

**Comment by tirumaraiselvan** [comment]
@CTOJoe We now have #384 implemented as well and available in `v1.2.1` onwards.
